extras/indexing/src/main/java/org/apache/rya/indexing/entity/EntityIndexSetProvider.java

```diff
@@ -0,0 +1,136 @@
+        subjectTypeMap = new HashMap<>();
+
+        //discover entities
+        final List<StatementPattern> unused = new ArrayList<>();
+        for (final QueryModelNode pattern : node.getOrderedNodes()) {
+            if(pattern instanceof StatementPattern) {
+                discoverEntities((StatementPattern) pattern, unused);
+            }
+        }
+
+        final List<EntityQueryNode> nodes = new ArrayList<>();
+        for(final Type type : typeMap.keySet()) {
+            //replace all nodes in the tupleExpr of the collection of statement patterns with this node.
+            final EntityQueryNode entity = new EntityQueryNode(type, typeMap.get(type), entityStorage);
+            nodes.add(entity);
+        }
+        return nodes;
+    }
+
+    private void discoverEntities(final StatementPattern pattern, final List<StatementPattern> unmatched) {
+        final Var subj = pattern.getSubjectVar();
+        final String subjStr = subj.getName();
+        final RyaURI predURI = getPredURI(pattern);
+        //check to see if current node is type
+        if(new URIImpl(predURI.getData()).equals(RDF.TYPE)) {
+            final Var obj = pattern.getObjectVar();
+            final RyaURI objURI = new RyaURI(obj.getValue().stringValue());
+            try {
+                final Optional<Type> optType = typeStorage.get(objURI);
+                //if is type, fetch type add to subject -> type map
+                if(optType.isPresent()) {
+                    final Type type = optType.get();
+                    typeMap.put(type, pattern);
+                    subjectTypeMap.put(subjStr, type);
+                    //check unmatched properties, add matches
+                    for(final StatementPattern propertyPattern : unmatched) {
+                        //store sp's as uri?
+                        final RyaURI property = getPredURI(propertyPattern);
+                        final Var typeSubVar = getTypeSubject(type);
+                        final Var patternSubVar = propertyPattern.getSubjectVar();
+                        if (type.getPropertyNames().contains(property) && typeSubVar.equals(patternSubVar)) {
+                            typeMap.put(type, propertyPattern);
+                        }
+                    }
+                }
+            } catch (final TypeStorageException e) {
+                e.printStackTrace();
+            }
+        } else {
+            //if not type, check to see if subject is in type map
+            if(subjectTypeMap.containsKey(subjStr)) {
+                //if is, check to see if pred is a property of type
+                final Type type = subjectTypeMap.get(subjStr);
+                if(type.getPropertyNames().contains(predURI)) {
+                    //if is, add sp to type -> sp map
+                    typeMap.put(type, pattern);
+                } else {
+                    //if not, add to unmatched type
+                    unmatched.add(pattern);
```

Comment:
I'm not so sure that this is what you want to do with a pattern whose subject name corresponds to a given type, but whose property doesn't match any of the type properties.  By adding it to unmatched you are allowing for the possibility that this pattern can be matched with another type corresponding to the same variable.  It seems as though you don't want to allow this given that subjectTypeMap is a hashmap (so you can associate only one type to each subject var name).  On the other hand, if you have are fed a segment that has two types associated with the same variable then your entry in subjectTypeMap will be overwritten and it is possible for you to have StatementPatterns assigned to each of the types in typeMap.  So I would add some sort of check to ensure you don't have multiple types associated with the same subject variable in your segment.  Once this is in place you don't need to add the above pattern to unmatched because there is no chance it will be matched to another type with the same variable.  With that in mind, you could probably revise the above branching structure to be if( ... equals(RDF.TYPE)) {...} else if(subjectTypeMap.containsKey(subjStr)) {...} else {unmatched.add(pattern)}